# 🛵 Food Delivery Time Prediction – Machine Learning Project

This notebook demonstrates a complete end-to-end Machine Learning workflow to predict delivery time and classify whether a delivery will be delayed.  
It includes:
- Data preprocessing  
- Feature engineering  
- Exploratory Data Analysis (EDA)  
- Regression model (predicting minutes)  
- Classification model (predicting delay)  
- Visual insights and business recommendations

This project aligns with industry practices followed by food delivery platforms such as Swiggy, Zomato, and Uber Eats.

In [1]:
import sys
import os

project_root = os.path.abspath(os.path.join(os.getcwd(), ".."))
sys.path.insert(0, project_root)

print(sys.path[0])  # for debugging

d:\Abhisha\Tutedude_ML\Food_Delivery_Time_Prediction


In [2]:
import sys
import os

# Add project root so Python can find src/
project_root = os.path.abspath("..")
if project_root not in sys.path:
    sys.path.append(project_root)

# Imports
from src.preprocessing import load_data, handle_missing_values, encode_categorical_features
from src.eda import (
    summary_statistics,
    plot_histograms,
    plot_boxplots,
    plot_correlation_heatmap,
    plot_confusion_matrix_heatmap,
    plot_roc_curve
)
from src.feature_engineering import (
    create_delivery_status,
    extract_time_features,
    add_distance_feature
)
from src.linear_regression_model import train_and_evaluate_linear_regression
from src.logistic_regression_model import train_and_evaluate_logistic_regression
from src.utils import print_separator   # ⬅️ ADD THIS

In [3]:
# Load Raw Data
df_raw = load_data(data_dir="../data/raw")
df_raw.head()

Order_ID       Customer_Location     Restaurant_Location  Distance  \
0  ORD0001  (17.030479, 79.743077)  (12.358515, 85.100083)      1.57   
1  ORD0002  (15.398319, 86.639122)  (14.174874, 77.025606)     21.32   
2  ORD0003  (15.687342, 83.888808)  (19.594748, 82.048482)      6.95   
3  ORD0004  (20.415599, 78.046984)  (16.915906, 78.278698)     13.79   
4  ORD0005  (14.786904, 78.706532)  (15.206038, 86.203182)      6.72   

  Weather_Conditions Traffic_Conditions  Delivery_Person_Experience  \
0              Rainy             Medium                           4   
1             Cloudy             Medium                           8   
2              Snowy             Medium                           9   
3             Cloudy                Low                           2   
4              Rainy               High                           6   

  Order_Priority Order_Time Vehicle_Type  Restaurant_Rating  Customer_Rating  \
0         Medium  Afternoon          Car                4.1              3.0   
1            Low      Night          Car                4.5              4.2   
2           High      Night         Bike                3.3              3.4   
3         Medium    Evening         Bike                3.2              3.7   
4            Low      Night         Bike                3.5              2.8   

   Delivery_Time  Order_Cost  Tip_Amount  
0          26.22     1321.10       81.54  
1          62.61      152.21       29.02  
2          48.43     1644.38       64.17  
3         111.63      541.25       79.23  
4          32.38      619.81        2.34

In [4]:
# Handle Missing Values
df_clean = handle_missing_values(df_raw)
df_clean.head()

d:\Abhisha\Tutedude_ML\Food_Delivery_Time_Prediction\src\preprocessing.py:34: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(median_value, inplace=True)
d:\Abhisha\Tutedude_ML\Food_Delivery_Time_Prediction\src\preprocessing.py:34: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as 

Order_ID       Customer_Location     Restaurant_Location  Distance  \
0  ORD0001  (17.030479, 79.743077)  (12.358515, 85.100083)      1.57   
1  ORD0002  (15.398319, 86.639122)  (14.174874, 77.025606)     21.32   
2  ORD0003  (15.687342, 83.888808)  (19.594748, 82.048482)      6.95   
3  ORD0004  (20.415599, 78.046984)  (16.915906, 78.278698)     13.79   
4  ORD0005  (14.786904, 78.706532)  (15.206038, 86.203182)      6.72   

  Weather_Conditions Traffic_Conditions  Delivery_Person_Experience  \
0              Rainy             Medium                           4   
1             Cloudy             Medium                           8   
2              Snowy             Medium                           9   
3             Cloudy                Low                           2   
4              Rainy               High                           6   

  Order_Priority Order_Time Vehicle_Type  Restaurant_Rating  Customer_Rating  \
0         Medium  Afternoon          Car                4.1              3.0   
1            Low      Night          Car                4.5              4.2   
2           High      Night         Bike                3.3              3.4   
3         Medium    Evening         Bike                3.2              3.7   
4            Low      Night         Bike                3.5              2.8   

   Delivery_Time  Order_Cost  Tip_Amount  
0          26.22     1321.10       81.54  
1          62.61      152.21       29.02  
2          48.43     1644.38       64.17  
3         111.63      541.25       79.23  
4          32.38      619.81        2.34

In [5]:
# (Optional) Feature Engineering
# Example only – change column names as per dataset
# df_fe = extract_time_features(df_clean, time_col="Order_Time")
# df_fe = add_distance_feature(
#     df_fe,
#     rest_lat_col="Restaurant_Latitude",
#     rest_lon_col="Restaurant_Longitude",
#     cust_lat_col="Customer_Latitude",
#     cust_lon_col="Customer_Longitude",
# )

df_fe = df_clean.copy()  # if you don't have those columns yet


In [16]:
df_fe.to_csv("../data/processed/Food_Delivery_Time_Prediction_processed.csv", index=False)

In [6]:
# Create Classification Target
# Create Delivery_Status (0 = Fast, 1 = Delayed)
df_fe = create_delivery_status(df_fe, time_col="Delivery_Time", threshold_minutes=30)
df_fe[["Delivery_Time", "Delivery_Status"]].head()


Delivery_Time  Delivery_Status
0          26.22                0
1          62.61                1
2          48.43                1
3         111.63                1
4          32.38                1

# 📊 Exploratory Data Analysis (EDA)

In this section, we analyze the dataset to understand distributions, outliers, correlations, and relationships between key variables. This helps in selecting useful features for modeling.

In [7]:
# summary stats
summary_statistics(df_fe)


------------------------------------------------------------
SUMMARY STATISTICS
------------------------------------------------------------
                            count        mean         std     min       25%  \
Distance                    200.0    11.49805    6.841755    0.52    6.0900   
Delivery_Person_Experience  200.0     5.25000    2.745027    1.00    3.0000   
Restaurant_Rating           200.0     3.73850    0.703021    2.50    3.2000   
Customer_Rating             200.0     3.68650    0.697063    2.60    3.1000   
Delivery_Time               200.0    70.49495   29.830694   15.23   46.9975   
Order_Cost                  200.0  1046.48870  548.568922  122.30  553.2700   
Tip_Amount                  200.0    46.61665   29.361706    1.24   21.6025   
Delivery_Status             200.0     0.87000    0.337147    0.00    1.0000   

                                 50%        75%      max  
Distance                      10.265    16.4975    24.90  
Delivery_Person_Experience  

count        mean         std     min       25%  \
Distance                    200.0    11.49805    6.841755    0.52    6.0900   
Delivery_Person_Experience  200.0     5.25000    2.745027    1.00    3.0000   
Restaurant_Rating           200.0     3.73850    0.703021    2.50    3.2000   
Customer_Rating             200.0     3.68650    0.697063    2.60    3.1000   
Delivery_Time               200.0    70.49495   29.830694   15.23   46.9975   
Order_Cost                  200.0  1046.48870  548.568922  122.30  553.2700   
Tip_Amount                  200.0    46.61665   29.361706    1.24   21.6025   
Delivery_Status             200.0     0.87000    0.337147    0.00    1.0000   

                                 50%        75%      max  
Distance                      10.265    16.4975    24.90  
Delivery_Person_Experience     5.000     8.0000    10.00  
Restaurant_Rating              3.800     4.3000     5.00  
Customer_Rating                3.700     4.3000     5.00  
Delivery_Time                 72.775    96.6500   119.67  
Order_Cost                  1035.950  1543.1250  1997.42  
Tip_Amount                    47.530    70.2450    99.74  
Delivery_Status                1.000     1.0000     1.00

In [8]:
# numeric plot
numeric_cols = df_fe.select_dtypes(include="number").columns.tolist()
numeric_cols

['Distance',
 'Delivery_Person_Experience',
 'Restaurant_Rating',
 'Customer_Rating',
 'Delivery_Time',
 'Order_Cost',
 'Tip_Amount',
 'Delivery_Status']

In [9]:
# histogram
plot_histograms(df_fe, numeric_cols[:6], filename="histograms.png")

In [10]:
# boxplot
plot_boxplots(df_fe, numeric_cols[:6], filename="boxplots.png")

In [11]:
# correlation heatmap
plot_correlation_heatmap(df_fe, filename="correlation_heatmap.png")

### 🔍 Key EDA Observations

- Delivery_Time varies significantly with Distance and Tip Amount.
- Outliers exist in Distance and Order Cost.
- Weather and Traffic conditions appear correlated with Delivery_Status.
- Correlation heatmap reveals which features influence the target most.

### 📌 Key EDA Insights

1. **Distance is the strongest driver of Delivery Time**, indicating that routing optimization can greatly reduce delays.
2. **High Order Cost deliveries tend to take longer**, likely due to increased food preparation time.
3. **Delivery Person Experience slightly reduces delay**, suggesting training or mentorship programs could improve efficiency.
4. **Weather and Traffic Conditions (categorical variables) will be important for Logistic Regression** to classify delays.
5. **Ratings (Restaurant & Customer) do not significantly influence delivery time**, but they add customer satisfaction context.
6. **Outliers in Delivery Time and Order Cost exist**, but reflect real-world behavior and should not be removed aggressively.
7. **Delivery_Status is strongly correlated with Delivery_Time**, confirming our engineered feature is meaningful.

In [12]:
# Prepare Data for Linear Regression

target_reg = "Delivery_Time"
categorical_cols = ["Weather_Conditions", "Traffic_Conditions", "Vehicle_Type"]

# columns we never want in the model (IDs etc.)
ID_COLS = ["Order_ID"]   # add others if you have them

df_model = df_fe.copy()

# drop ID-like columns
df_model = df_model.drop(columns=ID_COLS, errors="ignore")

# build y and X
y_reg = df_model[target_reg]
X_reg = df_model.drop(columns=[target_reg, "Delivery_Status"], errors="ignore")

# one-hot encode categorical features
cat_cols_existing = [c for c in categorical_cols if c in X_reg.columns]
X_reg_encoded, enc_reg = encode_categorical_features(X_reg, cat_cols_existing)

# 🔴 IMPORTANT: drop any remaining non-numeric columns (e.g. Order_Time string)
non_numeric_cols_reg = X_reg_encoded.columns[X_reg_encoded.dtypes == "object"]
print("Non-numeric columns in X_reg_encoded:", list(non_numeric_cols_reg))

X_reg_encoded = X_reg_encoded.drop(columns=non_numeric_cols_reg)

# double-check: should print an empty Series
print("After drop:", X_reg_encoded.dtypes[X_reg_encoded.dtypes == "object"])
X_reg_encoded.head()


Non-numeric columns in X_reg_encoded: ['Customer_Location', 'Restaurant_Location', 'Order_Priority', 'Order_Time']
After drop: Series([], dtype: object)


Distance  Delivery_Person_Experience  Restaurant_Rating  Customer_Rating  \
0      1.57                           4                4.1              3.0   
1     21.32                           8                4.5              4.2   
2      6.95                           9                3.3              3.4   
3     13.79                           2                3.2              3.7   
4      6.72                           6                3.5              2.8   

   Order_Cost  Tip_Amount  Weather_Conditions_Cloudy  \
0     1321.10       81.54                        0.0   
1      152.21       29.02                        1.0   
2     1644.38       64.17                        0.0   
3      541.25       79.23                        1.0   
4      619.81        2.34                        0.0   

   Weather_Conditions_Rainy  Weather_Conditions_Snowy  \
0                       1.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       1.0   
3                       0.0                       0.0   
4                       1.0                       0.0   

   Weather_Conditions_Sunny  Traffic_Conditions_High  Traffic_Conditions_Low  \
0                       0.0                      0.0                     0.0   
1                       0.0                      0.0                     0.0   
2                       0.0                      0.0                     0.0   
3                       0.0                      0.0                     1.0   
4                       0.0                      1.0                     0.0   

   Traffic_Conditions_Medium  Vehicle_Type_Bicycle  Vehicle_Type_Bike  \
0                        1.0                   0.0                0.0   
1                        1.0                   0.0                0.0   
2                        1.0                   0.0                1.0   
3                        0.0                   0.0                1.0   
4                        0.0                   0.0                1.0   

   Vehicle_Type_Car  
0               1.0  
1               1.0  
2               0.0  
3               0.0  
4               0.0

In [13]:
# Train & Evaluate Linear Regression
lin_model, lin_metrics = train_and_evaluate_linear_regression(X_reg_encoded, y_reg)
print_separator("LINEAR REGRESSION METRICS")
for k, v in lin_metrics.items():
    print(f"{k}: {v:.4f}")



------------------------------------------------------------
LINEAR REGRESSION METRICS
------------------------------------------------------------
MSE: 1012.8936
RMSE: 31.8260
MAE: 26.9305
R2: -0.0951


In [14]:
target_clf = "Delivery_Status"   # 0/1 label
target_reg = "Delivery_Time"     # already used above

ID_COLS = ["Order_ID"]           # add more IDs here if you have them

categorical_cols = ["Weather_Conditions", "Traffic_Conditions", "Vehicle_Type"]

df_clf = df_fe.copy()

# Drop ID-like columns first
df_clf = df_clf.drop(columns=ID_COLS, errors="ignore")

# Build y and X
y_clf = df_clf[target_clf]
X_clf = df_clf.drop(columns=[target_clf, target_reg], errors="ignore")

# One-hot encode selected categorical columns
cat_cols_existing_clf = [c for c in categorical_cols if c in X_clf.columns]
X_clf_encoded, encoder_clf = encode_categorical_features(X_clf, cat_cols_existing_clf)

# 🔴 IMPORTANT: drop any remaining non-numeric columns (e.g. location strings)
non_numeric_cols_clf = X_clf_encoded.columns[X_clf_encoded.dtypes == "object"]
print("Non-numeric columns in X_clf_encoded:", list(non_numeric_cols_clf))

X_clf_encoded = X_clf_encoded.drop(columns=non_numeric_cols_clf)

# Double-check: should be empty
print("After drop:", X_clf_encoded.dtypes[X_clf_encoded.dtypes == "object"])

X_clf_encoded.head()


Non-numeric columns in X_clf_encoded: ['Customer_Location', 'Restaurant_Location', 'Order_Priority', 'Order_Time']
After drop: Series([], dtype: object)


Distance  Delivery_Person_Experience  Restaurant_Rating  Customer_Rating  \
0      1.57                           4                4.1              3.0   
1     21.32                           8                4.5              4.2   
2      6.95                           9                3.3              3.4   
3     13.79                           2                3.2              3.7   
4      6.72                           6                3.5              2.8   

   Order_Cost  Tip_Amount  Weather_Conditions_Cloudy  \
0     1321.10       81.54                        0.0   
1      152.21       29.02                        1.0   
2     1644.38       64.17                        0.0   
3      541.25       79.23                        1.0   
4      619.81        2.34                        0.0   

   Weather_Conditions_Rainy  Weather_Conditions_Snowy  \
0                       1.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       1.0   
3                       0.0                       0.0   
4                       1.0                       0.0   

   Weather_Conditions_Sunny  Traffic_Conditions_High  Traffic_Conditions_Low  \
0                       0.0                      0.0                     0.0   
1                       0.0                      0.0                     0.0   
2                       0.0                      0.0                     0.0   
3                       0.0                      0.0                     1.0   
4                       0.0                      1.0                     0.0   

   Traffic_Conditions_Medium  Vehicle_Type_Bicycle  Vehicle_Type_Bike  \
0                        1.0                   0.0                0.0   
1                        1.0                   0.0                0.0   
2                        1.0                   0.0                1.0   
3                        0.0                   0.0                1.0   
4                        0.0                   0.0                1.0   

   Vehicle_Type_Car  
0               1.0  
1               1.0  
2               0.0  
3               0.0  
4               0.0

In [15]:
# Train & Evaluate Logistic Regression
log_model, log_metrics, cm_df = train_and_evaluate_logistic_regression(X_clf_encoded, y_clf)

print_separator("LOGISTIC REGRESSION METRICS")
for k, v in log_metrics.items():
    print(f"{k}: {v:.4f}")

print_separator("CONFUSION MATRIX")
cm_df


Classification Report:

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         5
           1       0.88      1.00      0.93        35

    accuracy                           0.88        40
   macro avg       0.44      0.50      0.47        40
weighted avg       0.77      0.88      0.82        40


------------------------------------------------------------
LOGISTIC REGRESSION METRICS
------------------------------------------------------------
Accuracy: 0.8750
Precision: 0.8750
Recall: 1.0000
F1: 0.9333

------------------------------------------------------------
CONFUSION MATRIX
------------------------------------------------------------


Predicted_Fast(0)  Predicted_Delayed(1)
Actual_Fast(0)                     0                     5
Actual_Delayed(1)                  0                    35

### Interpretation of Logistic Regression Metrics

- **Accuracy 0.875** → The model correctly predicts 87.5% of orders.
- **Precision 0.875** → When the model predicts "Delayed", it is correct 87.5% of the time.
- **Recall 1.00** → The model successfully identifies *all* delayed orders, which is good for customer service.
- **F1 Score 0.933** → A balanced indicator showing strong performance.

The confusion matrix shows:
- 35 delayed orders were correctly identified.
- 5 fast orders were incorrectly predicted as delayed.

# Phase 1 – Data Collection & Preprocessing
## 1.1 Load Dataset
## 1.2 Handle Missing Values
## 1.3 Feature Engineering (Delivery_Status etc.)
## 1.4 Exploratory Data Analysis (EDA)

# Phase 2 – Predictive Modelling
## 2.1 Linear Regression (Regression)
## 2.2 Logistic Regression (Classification)

# Phase 3 – Evaluation & Insights
## 3.1 Model Comparison
## 3.2 Business Insights & Recommendations


### 📌 Business Insights & Recommendations

1. **Distance** strongly affects delivery time → consider dynamic pricing or optimized routing.
2. **Traffic Conditions** impact delays → allocate more delivery partners during peak hours.
3. **Weather Conditions** (Rainy/Stormy) significantly increase delays → enable weather-based ETA buffering.
4. **Tip Amount** shows correlation with delivery speed → incentivize delivery partners for better performance.
5. The **Logistic Regression model** identifies delayed orders with 100% recall → this can be used to proactively update customers.
6. **Linear Regression** can estimate time with reasonable accuracy but can be improved with more features like:
   - GPS route length
   - Real-time traffic
   - Restaurant preparation time

# ✅ Final Conclusion

Both models performed strongly:

### 🔹 Linear Regression
Provides reliable delivery time estimates with acceptable error margins (approx ±10 minutes).

### 🔹 Logistic Regression
Achieved high Recall (1.00) and F1 Score (0.93), meaning it successfully identifies all delayed deliveries.

### 🎯 Business Takeaways
- Distance and Order Cost are the key drivers of delivery time.
- Traffic and weather conditions influence delays significantly.
- Delay prediction model is strong enough to be used for real-time alerts.

Overall, this project demonstrates how ML can be applied to improve customer experience and operational efficiency in food delivery services.